### PR TITLE
kvserver: improve Raft snapshot log messages

### LINKED ISSUE #48172: kv: snapshot rcv/send should log the node ID of the other side
Found by @irfansharif on https://github.com/cockroachlabs/support/issues/425

Today we have two messages on both sides of a snapshot:

```
I200321 10:04:47.078260 828716517 storage/store_snapshot.go:688  [n5,raftsnapshot,s5,r4/22:/System/{NodeLive…-tsd}] sending Raft snapshot 14fa20da at applied index 19542277
```

and

```
I200404 10:03:53.839268 7359465266 storage/replica_raftstorage.go:822  [n4,s4,r4/14:/System/{NodeLive…-tsd}] applying Raft snapshot at index 24781485 (id=2c1ce645, encoded size=306796, 1 rocksdb batches, 5 log entries)
I200404 10:03:53.848034 7359465266 storage/replica_raftstorage.go:828  [n4,s4,r4/14:/System/{NodeLive…-tsd}] applied Raft snapshot in 9ms [clear=5ms batch=0ms entries=0ms commit=2ms]
```

However the former does not say where the snapshot is sent *to*

and the latter does not say where the snapshot is coming *from*

Both should be logged.

### PR DESCRIPTION
**Pull request by erikgrinaker**
Previously, Raft snapshot application log messages did not include the
sending replica, and had lots of redaction markers:

```
[n1,replicate,s1,r42/1:‹/Table/4{6-7}›] 552 streamed INITIAL snapshot ‹90e5261f› at applied index 22 to (n2,s2):3LEARNER in 0.00s @ ‹15 MiB›/s: ‹kv pairs: 13›, rate-limit: ‹32 MiB›/s, queued: 0.00s
[n2,s2,r42/3:{-}] 175 applying snapshot of type INITIAL [id=‹90e5261f› index=22]
[n2,s2,r42/3:‹/Table/4{6-7}›] 176 applied snapshot of type INITIAL [‹total=17ms ›‹ingestion=6@13ms ›‹›id=‹90e5261f› index=22]
```

This patch includes sender information, fixes the redaction markers, and
delegates to the snapshot's `String` method:

```
[n1,replicate,s1,r42/1:‹/Table/4{6-7}›] 552 streamed INITIAL snapshot 90e5261f at applied index 22 to (n2,s2):3LEARNER in 0.00s @ 15 MiB/s: kv pairs: 13, rate-limit: 32 MiB/s, queued: 0.00s
[n2,s2,r42/3:{-}] 175 applying INITIAL snapshot 90e5261f from (n1,s1):1 at applied index 22
[n2,s2,r42/3:‹/Table/4{6-7}›] 176 applied INITIAL snapshot 90e5261f from (n1,s1):1 at applied index 22 (total=17ms ingestion=6@13ms)
```
A new field `IncomingSnapshot.FromReplica` is added to thread the
sender information through the snapshot application code path.

Resolves #48172.

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_raftstorage.go [hunk 1]
```diff
@@ -809,31 +812,25 @@ func (r *Replica) applySnapshot(
 		// Time to ingest SSTs.
 		ingestion time.Time
 	}
-	log.Infof(ctx, "applying snapshot of type %s [id=%s index=%d]", inSnap.snapType,
-		inSnap.SnapUUID.Short(), nonemptySnap.Metadata.Index)
+	log.Infof(ctx, "applying snapshot of type %s [id=%s from=%s index=%d]", inSnap.snapType,
```
Comment (tbg) [inline pkg/kv/kvserver/replica_raftstorage.go:815]: shouldn't this ideally be something like `log.Infof(ctx, applying %s, inSnap)` (which uses `(*IncomingSnapshot).SafeFormat`?

File: pkg/kv/kvserver/replica_raftstorage.go [hunk 2]
```diff
@@ -809,31 +812,25 @@ func (r *Replica) applySnapshot(
...
+			" ingestion=%d@%0.0fms",
 			len(inSnap.SSTStorageScratch.SSTs()),
-			stats.ingestion.Sub(stats.subsumedReplicas).Seconds()*1000,
-		)
-		log.Infof(
-			ctx, "applied snapshot of type %s [%s%s%sid=%s index=%d]", inSnap.snapType, totalLog,
-			subsumedReplicasLog, ingestionLog, inSnap.SnapUUID.Short(), nonemptySnap.Metadata.Index,
-		)
+			stats.ingestion.Sub(stats.subsumedReplicas).Seconds()*1000)
+		log.Infof(ctx, "applied snapshot of type %s [id=%s from=%s index=%d%s%s%s]", inSnap.snapType,
```
Comment (tbg) [inline pkg/kv/kvserver/replica_raftstorage.go:831]: ditto, `log.Infof(ctx, "applied %s%s%s%s", inSnap, totalLog, subsumedReplicasLog, ingestionLog)`

File: pkg/kv/kvserver/store_snapshot.go [hunk 3]
```diff
@@ -374,7 +376,9 @@ func (kvSS *kvBatchSnapshotStrategy) sendBatch(
 }
 
 // Status implements the snapshotStrategy interface.
-func (kvSS *kvBatchSnapshotStrategy) Status() string { return kvSS.status }
+func (kvSS *kvBatchSnapshotStrategy) Status() redact.SafeString {
+	return redact.SafeString(kvSS.status)
```
Comment (tbg) [inline pkg/kv/kvserver/store_snapshot.go:380]: would be cleaner to make the `status` field a `redact.SafeString`, no?

File: pkg/kv/kvserver/store_snapshot.go [hunk 4]
```diff
@@ -970,9 +974,9 @@ func sendSnapshot(
 		snap,
 		to,
 		durSent.Seconds(),
-		humanizeutil.IBytes(int64(float64(numBytesSent)/durSent.Seconds())),
+		redact.Safe(humanizeutil.IBytes(int64(float64(numBytesSent)/durSent.Seconds()))),
```
Comment (tbg) [inline pkg/kv/kvserver/store_snapshot.go:977]: if we make `IBytes` return a `SafeString` as we could, I assume the diff blows up wildly, right?

**Review by tbg — APPROVED** [review]

File: pkg/kv/kvserver/replica_raftstorage.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_raftstorage.go:815]: Notice that the `index` field here comes from `nonemptySnap`, not `inSnap`.

File: pkg/kv/kvserver/store_snapshot.go (hunk 4 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/store_snapshot.go:977]: Yeah, started off doing that, but it got pretty involved. I'll submit a follow-up PR.

File: pkg/kv/kvserver/store_snapshot.go (hunk 3 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/store_snapshot.go]: True, done.

File: pkg/kv/kvserver/replica_raftstorage.go (hunk 1 above)
Comment (tbg) [reply pkg/kv/kvserver/replica_raftstorage.go:815]: All of the various index fields must be the same.

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_raftstorage.go]: Ok, thanks -- updated.

**Review by tbg — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @knz)

___
*[pkg/kv/kvserver/replica_raftstorage.go, line 817 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72684#-MoZ5D6G3ryFZlW2Prj8:-MoZ5D6G3ryFZlW2Prj9:bkmmc8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d7fb5ccf02d2f2cb798a07a452c573aac735b12/pkg/kv/kvserver/replica_raftstorage.go#L817)):*
> ```Go
> 	log.Infof(ctx, "applying %s", inSnap)
> 	defer func(start time.Time) {
> 		logDetails := redact.Sprintf("total=%0.0fms", timeutil.Now().Sub(start).Seconds()*1000)
> ```

Ah man, now you're really close to just using a `redact.StringBuilder`, I just have to suggest it. Sorry :)

**Review by erikgrinaker — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/kv/kvserver/replica_raftstorage.go, line 817 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72684#-MoZ5D6G3ryFZlW2Prj8:-MoZE3O-7P4KVclpDofb:b6jxeyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d7fb5ccf02d2f2cb798a07a452c573aac735b12/pkg/kv/kvserver/replica_raftstorage.go#L817)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ah man, now you're really close to just using a `redact.StringBuilder`, I just have to suggest it. Sorry :)
</blockquote></details>

Hehe sure, done.

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

**Comment by erikgrinaker** [comment]
TFTR!

bors r=tbg

**Pull request merged**